solo-io/gloo
# Discovery resource version err

# Description

Fix a bug that caused Resync for discovered Upstreams to fail when `watchLabels` were in use

## Code changes
- Remove watchLabels from the selector used to select Upstreams for Resync
  - Add kube2e test 

# Context

See #8635 for context on issue, as well as [root cause](https://github.com/solo-io/gloo/issues/8635#issuecomment-2374785559) and [reasoning around the solution](https://github.com/solo-io/gloo/issues/8635#issuecomment-2374923454)

I did encounter #8227 as well while working on this but have not investigated that piece further

This bug has been present since at least v1.14 and therefore we expect to backport this to all LTS branches

## Testing steps

### Manual testing

I followed the reproduction steps from the issue to reproduce the bug, observing both the error in logs and that the discovered Upstream would not get updates when its parent Service was updated

I built the discovery image from this branch, loaded it into KinD, and modified the `discovery` deployment to use the test image
```
$ VERSION=0.0.0-test make -B kind-build-and-load-discovery
```

Following the same reproduction steps I then observed that the Upstream was, in fact, updated when its parent service was, and that logs no longer contained errors

### Automated testing

There is no testing infrastructure for unit tests of Discovery resync, and any such tests would require extensive mocking of k8s, complicated async setup, etc. and would therefore represent a significant lift
In my estimation this would not be worthwhile as a kube2e test, which has been added, is sufficient to test the buggy behavior

# Checklist:
